Modules/_ctypes/cfield.c

```diff
@@ -147,61 +233,34 @@ PyCField_FromDesc(PyObject *desc, Py_ssize_t index,
     Py_INCREF(proto);
     self->proto = proto;
 
-    switch (fieldtype) {
-    case NEW_BITFIELD:
-        if (big_endian)
-            self->size = (bitsize << 16) + *pfield_size - *pbitofs - bitsize;
-        else
-            self->size = (bitsize << 16) + *pbitofs;
-        *pbitofs = bitsize;
-        /* fall through */
-    case NO_BITFIELD:
-        if (pack)
-            align = min(pack, dict->align);
-        else
-            align = dict->align;
-        if (align && *poffset % align) {
-            Py_ssize_t delta = align - (*poffset % align);
-            *psize += delta;
-            *poffset += delta;
-        }
-
-        if (bitsize == 0)
-            self->size = size;
-        *psize += size;
-
-        self->offset = *poffset;
-        *poffset += size;
-
-        *palign = align;
-        break;
-
-    case EXPAND_BITFIELD:
-        *poffset += dict->size - *pfield_size/8;
-        *psize += dict->size - *pfield_size/8;
-
-        *pfield_size = dict->size * 8;
-
-        if (big_endian)
-            self->size = (bitsize << 16) + *pfield_size - *pbitofs - bitsize;
-        else
-            self->size = (bitsize << 16) + *pbitofs;
-
-        self->offset = *poffset - size; /* poffset is already updated for the NEXT field */
-        *pbitofs += bitsize;
-        break;
-
-    case CONT_BITFIELD:
-        if (big_endian)
-            self->size = (bitsize << 16) + *pfield_size - *pbitofs - bitsize;
-        else
-            self->size = (bitsize << 16) + *pbitofs;
-
-        self->offset = *poffset - size; /* poffset is already updated for the NEXT field */
-        *pbitofs += bitsize;
-        break;
+    int is_bitfield = !!bitsize;
+    if(!is_bitfield) {
+        bitsize = 8 * dict->size;
+        // Caution: bitsize might still be 0 now.
+    }
+    assert(bitsize <= dict->size * 8);
+
+    // `pack` only makes sense in msvc compatibility mode.
```

Comment:
https://docs.python.org/3/library/ctypes.html#structure-union-alignment-and-byte-order says

> By default, Structure and Union fields are aligned in the same way the C compiler does it. It is possible to override this behavior by specifying a _pack_ class attribute in the subclass definition. This must be set to a positive integer and specifies the maximum alignment for the fields. This is what #pragma pack(n) also does in MSVC.

GCC has a suspiciously similarly named `__attribute__((packed))` / `#pragma pack`.  But it does something different.  To get GCC to give you the behaviour as described in the doc, you also need to enable `__attribute__((ms_struct))` / `#pragma ms_struct on`.  And that's what we are doing here.

Reply:
Note: the behaviour described in the docs that we are implementing here is rather different from what everyone in https://github.com/python/cpython/pull/19850 seems to have understood.

As far as I can tell in that PR everyone assume that `_packed_ = 1` means GCC's `__attribute__((packed))`.

This misunderstanding might be an argument for changing the docs?  However that would be a breaking change.

In general, this PR breaks bug-for-bug compatibility with the past, but does not change documented behaviour.

Reply:
Correction: support for GCC's packing isn't actually as easy as I imagined.

The problem is that GCC's packing can make eg a `uint8_t` straddle two bytes.  In order to support that, we'd need to revamp much more of ctypes.

Reply:
Here's an example to show the problem:

```c
#include <stdio.h>
#include <stdint.h>
#include <string.h>

typedef struct
__attribute__ ((packed))
{
  uint8_t a: 4;
  uint8_t b: 8;
  uint8_t c: 4;
} Foo;

typedef union {
  Foo foo;
  uint8_t x[sizeof(Foo)];
} U;

int main(int argc, char** argv) {
    printf("%lu\n", __alignof__(Foo));
    printf("%lu\n", sizeof(Foo));

    U u;
    memset(&u, 0, sizeof(U));
    u.foo.b = -1;
    // Because we are on a little endian machine, it makes more sense to print
    // the bytes in reverse order.
    // We expect this to print: "0f f0"
    for(int i=sizeof(Foo)-1; i>=0; --i) {
        printf("%02x ", u.x[i]);
    }
    printf("\n");
    return 0;
}
```

Reply:
I guess the right thing to do here would be to warn when `_packed_` is used with the GCC semantics.